numpy/_core/src/multiarray/iterators.c

```diff
@@ -561,195 +489,154 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
             if (cast_info.func(&cast_info.context, args, &one,
                                transfer_strides, cast_info.auxdata) < 0) {
-                goto fail;
+                goto finish;
             }
-            start += step_size;
+            start += step;
             PyArray_ITER_GOTO1D(self, start);
-            dptr += size;
+            dptr += dtype_size;
         }
         PyArray_ITER_RESET(self);
-        NPY_cast_info_xfree(&cast_info);
-        return (PyObject *)ret;
+        goto finish;
     }
 
-    /* convert to INTP array if Integer array scalar or List */
-    indtype = PyArray_DescrFromType(NPY_INTP);
-    if (PyArray_IsScalar(ind, Integer) || PyList_Check(ind)) {
-        Py_INCREF(indtype);
-        obj = PyArray_FromAny(ind, indtype, 0, 0, NPY_ARRAY_FORCECAST, NULL);
-        if (obj == NULL) {
-            goto fail;
-        }
-    }
-    else {
-        Py_INCREF(ind);
-        obj = ind;
+    if (index_type == HAS_BOOL) {
+        ret = iter_subscript_Bool(self, (PyArrayObject *) indices[0].object, &cast_info);
+        goto finish;
     }
 
-    if (!PyArray_Check(obj)) {
-        PyArrayObject *tmp_arr = (PyArrayObject *) PyArray_FROM_O(obj);
-        if (tmp_arr == NULL) {
-            goto fail;
-        }
-
-        if (PyArray_SIZE(tmp_arr) == 0) {
-            PyArray_Descr *indtype = PyArray_DescrFromType(NPY_INTP);
-            Py_SETREF(obj, PyArray_FromArray(tmp_arr, indtype, NPY_ARRAY_FORCECAST));
-            Py_DECREF(tmp_arr);
-            if (obj == NULL) {
-                goto fail;
-            }
-        }
-        else {
-            Py_SETREF(obj, (PyObject *) tmp_arr);
+    if (index_type == HAS_FANCY) {
+        PyArray_Descr *indtype = PyArray_DescrFromType(NPY_INTP);
+        PyArrayObject *cast_array = (PyArrayObject *)
+            PyArray_FromArray((PyArrayObject *) indices[0].object, indtype, NPY_ARRAY_FORCECAST);
+        if (cast_array == NULL) {
+            goto finish;
         }
-    }
-
-    /* Check for Boolean array */
-    if (PyArray_TYPE((PyArrayObject *)obj) == NPY_BOOL) {
-        ret = iter_subscript_Bool(self, (PyArrayObject *)obj, &cast_info);
+        ret = iter_subscript_int(self, cast_array, &cast_info);
+        Py_DECREF(cast_array);
         goto finish;
     }
 
-    /* Only integer arrays left */
-    if (!PyArray_ISINTEGER((PyArrayObject *)obj)) {
-        goto fail;
-    }
-
-    Py_INCREF(indtype);
-    new = PyArray_FromAny(obj, indtype, 0, 0,
-                      NPY_ARRAY_FORCECAST | NPY_ARRAY_ALIGNED, NULL);
-    if (new == NULL) {
-        goto fail;
-    }
-    ret = (PyArrayObject *)iter_subscript_int(self, (PyArrayObject *)new,
-                                              &cast_info);
-    Py_DECREF(new);
-
- finish:
-    Py_DECREF(indtype);
-    Py_DECREF(obj);
+    PyErr_SetString(PyExc_IndexError,
+        "only integers, slices (`:`), ellipsis (`...`) and integer or boolean "
+        "arrays are valid indices"
```

Comment:
Ah, I suspect this can't be reached, but it seems good to have as a back-stop.  I'll just apply the `==` above and hope it works fine.

(Just to be a bit pedantic.)